gremlin-dotnet/src/Gremlin.Net/Process/Traversal/Bytecode.cs

```diff
@@ -79,7 +85,77 @@ public void AddSource(string sourceName, params object[] args)
         /// <param name="args">The traversal method arguments.</param>
         public void AddStep(string stepName, params object[] args)
         {
-            StepInstructions.Add(new Instruction(stepName, args));
+            StepInstructions.Add(new Instruction(stepName, FlattenArguments(args)));
+            Bindings.Clear();
```

Comment:
Even though `Bindings` is based on a `ThreadLocal<T>` instance, I think the implementation is not thread-safe.

For example: Multiple tasks executed serially on the same thread, modifying the same bindings dictionary.
Besides not being thread-safe, it doesn't support defining a binding on 1 thread and adding the step on another, sample:

```csharp
// thread 1
// define bindings
Bindings.Instance.Of(/* */);
await SomeUnRelatedTask();
// thread 2
// Do something with that binding
g.V()/**/;
```

Reply:
This is definitely a problem, but do you see a way to fix it? Unfortunately, I think we have to use `static` here unless we want to create overloads for each parameter in the Traversal API that takes a `Binding` instead of the unbound variable which would result in a lot of overloads.

Since [`Bindings` don't seem to improve the performance with Bytecode](https://lists.apache.org/thread.html/c10baff72e67a8b7728c42b5c4e983a83e82b8f6964b3aa465f0e341@<dev.tinkerpop.apache.org>) when no lambdas are used which we don't support anyway, we may also remove the support for `Bindings` altogether from Gremlin.Net. Then we also don't need the `FlattenArguments` and `ConvertArgument` functions in `Bytecode` which are not exactly intuitive.

Or if we want to continue supporting them, we could also add a note to the documentation to make users aware of these concurrency problems with `Bindings`.

Reply:
I agree that a solution could be an overkill on the API...

I'm +1 from not supporting bindings in the .NET GLV (and document it), I don't see much purpose.

What do you think about not supporting Bindings on the .NET GLV @okram @spmallette @robertdale ?

Reply:
Any reason we don't support lambdas? Even if .NET can't support them natively for some reason wouldn't we minimally support the ability to pass a python/groovy/etc lambda? it's kinda weird that way, but i think back to the point that kuppitz made on the dev list the other day where he stated that he doesn't always find a way out of using lambdas in production systems he works on - so ultimately users will need that kind of capability i think.

Reply:
That's a good point. So we should probably support lambdas. To be honest, I haven't really looked into what would be necessary to support lambdas. If I'm not mistaken then we can easily support Python and Groovy lambdas when we add a dedicated GraphSON serializer for them and probably a class that wraps the lambda together with its language. Then the steps that take a lambda would simply accept an object of that class in Gremlin.Net. I can create a separate ticket for this when we agree that we want to support them.

When we want to support lambdas then we should also support `Bindings`. That leaves us with the question of whether we find a better implementation that doesn't suffer from the concurrency problems @jorgebay mentioned or if the current implementation in this pull request is acceptable.

Apart from that, shouldn't we clarify the section in the documentation about bindings? It currently states for example [for gremlin-python](http://tinkerpop.apache.org/docs/current/reference/#_bindings):

> If a traversal is going to be executed repeatedly, but with different parameters, then bindings should be used.

because:
> translation and compilation can be skipped as the previously compiled version should be cached

which is only true when a lambda is used if I understood it correctly.